Expose mappings of requirements, provides and modules to modules

hikinggrass opened the pull request:
EVerest modules and even individual interface implementations can have mappings
assigned to them. These mappings are inspired by the OCPP 3-tier model.

Modules can access the mapping information in the following ways depending on which specific information is required, all of these return an optional `Mapping` struct
If the mapping of a requirement is of interest it can be accessed via a get_mapping() function of a requirments:
`r_name_of_the_requirement->get_mapping()`

If the mapping of an interface implementation is of interest it can also be accessed via a get_mapping() function: `p_name_of_an_implementation->get_mapping()`

If the mapping of the current module is of interest it can be accessed via the module info:
`this->info.mapping`

Mapping information is also available in error reporting via "error.origin.mapping":
```
const auto error_handler = [this](const Everest::error::Error& error) {
    const auto evse_id = error.origin.mapping.has_value() ? error.origin.mapping.value().evse : 0;
};
```

hikinggrass committed:
WIP: make mappings available outside of errors

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add stream operators for (optional) Mapping for easier debugging

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add module mapping to module info

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Bump version to 0.18

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

Pietfried reviewed (approved):
lgtm. Linting is currently failing

Tested with https://github.com/EVerest/everest-core/pull/872

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/expose-mappings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 commented on include/utils/types.hpp:
```diff
@@ -96,6 +76,30 @@ struct Mapping {
+    if (mapping.has_value()) {
+        os << mapping.value();
+    } else {
+        os << "Mapping(charging station)";
```
Does this default representation make any sense?

hikinggrass replied:
The feature (at the moment at least) is inspired by the 3-tier mappings from OCPP. So if you have no explicit mapping it's assumed that the module maps to the whole charging station, getting more specific by a mapping to an evse and/or connector

a-w50 replied:
I see.  Still I'm unsure, whether the string `Mapping(charging station)` on a text console will help users to understand what is meant.

a-w50 commented on include/utils/types.hpp:
```diff
@@ -96,6 +76,30 @@ struct Mapping {
+
+/// \brief Writes the string representation of the given Mapping \p mapping to the given output stream \p os
+/// \returns an output stream with the Mapping written to
+inline std::ostream& operator<<(std::ostream& os, const std::optional<Mapping>& mapping) {
```
The stream output operator should probably be defined for `Mapping` and not `std::optional<Mapping>`

hikinggrass replied:
It already is: https://github.com/EVerest/everest-framework/blob/557932c30108ce09b3670f01cb49c5d5092b681f/include/utils/types.hpp#L81

a-w50 commented on include/utils/types.hpp:
```diff
@@ -104,11 +108,39 @@ struct ModuleTierMappings {
+    std::optional<Mapping> mapping;
+};
+
+struct RequirementConnection {
```
What is a `RequirementConnection`?  Maybe there's a better word for this.  Maybe it is a fulfillment?

hikinggrass replied:
Renamed to Fulfillment as already used by everestpy

a-w50 reviewed (changes requested):
I would like to request some documentation in the commit message about what has been implemented here.

hikinggrass committed:
Rename RequirementConnection to Fulfillment

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 commented on include/framework/ModuleAdapter.hpp:
```diff
@@ -107,6 +109,8 @@ struct ModuleAdapter {
     std::vector<cmd> registered_commands;
     TelemetryPublishFunc telemetry_publish;
+    GetMappingFunc get_mapping;
+    GetImplMappingFunc get_impl_mapping;
```
Is it possible to have only one function instead of two, which then would return a struct with the module mappings and implementation mappings?

a-w50 replied:
Btw., we should discuss, if we really need all these `std::functions` here, it might be simpler to have the `ModuleAdapter` just having regular functions and its own private reference to the `everest` instance, and then the functions are implemented by just calling the corresponding functions on the `everest` instance.

hikinggrass replied:
Changed into one function, I consider refactoring this out-of-scope for this PR

a-w50 commented on include/framework/everest.hpp:
```diff
@@ -142,6 +142,16 @@ class Everest {
+    ///
+    /// \returns the 3 tier model mapping for the given \p impl_id
+    ///
+    std::optional<Mapping> get_3_tier_model_mapping(const std::string& impl_id);
```
In line with the Liskov Substitution Principle, this choice of api naming is maybe a bit unfortunate.  Both methods are named the same, but return different types of objects.  Furthermore they are not `const` correct.
I would advice to only have one accessor method (`const std::optional<ModuleTierMappings>& get_3_tier_model_mapping() const`, note: I would also return by const ref instead of copying).  Additionally you can implement a free helper function like `std::optional<Mapping> get_implementation_mapping(const ModuleTierMapping&, const std::string& impl_id)` which would do the same as `std::optional<Mapping> get_3_tier_model_mapping(const std::string& impl_id)` currently does.  This keeps the interface of the `Everest` class small and also allows you to easily test the function `get_implementation_mapping`.

a-w50 commented on include/utils/config.hpp:
```diff
@@ -147,9 +147,9 @@ class Config {
 
     ///
-    /// \returns a list of Requirements for \p module_id
+    /// \returns a map of Fulfillments for \p module_id
```
I really doubt the value of these comments, you get exactly the same information if you look at the function signature.

a-w50 commented on include/utils/config.hpp:
```diff
@@ -147,9 +147,9 @@ class Config {
+    /// \returns a map of Fulfillments for \p module_id
     ///
-    std::list<Requirement> get_requirements(const std::string& module_id) const;
+    std::map<std::string, std::vector<Fulfillment>> get_fulfillments(const std::string& module_id);
```
why isn't this function `const` anymore?

hikinggrass replied:
well spotted, re-added the const

a-w50 commented on include/utils/types.hpp:
```diff
@@ -104,11 +108,41 @@ struct ModuleTierMappings {
-    size_t index;
+    size_t index = 0;
+    std::optional<Mapping> mapping;
+};
```
Side note: this should just be a simple aggregate.  The `operator<` should be implemented as free function

a-w50 commented on include/utils/types.hpp:
```diff
@@ -104,11 +108,41 @@ struct ModuleTierMappings {
+    std::optional<Mapping> mapping;
+};
+
+/// \brief Describes a how a requirement is fulfilled by a connection to another module and its implementation
```
This doc comment should be more like:
A fulfillment relates a requirement to its connected implementation, identified via its module and implementation id.

hikinggrass replied:
I've applied your suggestion

a-w50 commented on include/utils/types.hpp:
```diff
@@ -104,11 +108,41 @@ struct ModuleTierMappings {
     std::string id;
-    size_t index;
+    size_t index = 0;
+    std::optional<Mapping> mapping;
```
I wouldn't add the `mapping` into the `Requirement` struct. The requirement was used as an identifier previously.  Of course it is very convenient, but you could equally add the fulfillments to this struct and so on and so on.  It is probably better to have a separate lookup struct/map/list with Requirement -> Mapping

hikinggrass replied:
Removed from Requirement (and added to Fulfillment)

a-w50 commented on lib/config.cpp:
```diff
@@ -705,23 +705,31 @@ json Config::resolve_requirement(const std::string& module_id, const std::string
 }
 
-std::list<Requirement> Config::get_requirements(const std::string& module_id) const {
+std::map<std::string, std::vector<Fulfillment>> Config::get_fulfillments(const std::string& module_id) {
```
As mentioned above, I would not add `mapping` to the `Requirement` struct, and hence this function and also the functions using this one, would just stay the same.  Better create the above mentioned struct (which contains the mapping `Requirement/Module -> mapping`) either on the fly or precompute it and then just return that.

a-w50 commented on lib/config.cpp:
```diff
@@ -823,7 +831,8 @@ std::optional<Mapping> Config::get_3_tier_model_mapping(const std::string& modul
     }
     auto& mapping = module_tier_mappings.value();
```
This code has been added before, but please use `const` where possible.

a-w50 commented on lib/config.cpp:
```diff
@@ -823,7 +831,8 @@ std::optional<Mapping> Config::get_3_tier_model_mapping(const std::string& modul
     }
     auto& mapping = module_tier_mappings.value();
     if (mapping.implementations.find(impl_id) == mapping.implementations.end()) {
-        return std::nullopt;
+        // if no specific implementation mapping is given, use the module mapping
+        return mapping.module;
     }
     return mapping.implementations.at(impl_id);
```
This pattern happens quite often.  It is usually better to use the iterator instead of doing the lookup twice:
```
const auto mapping_it = ...find(impl_id);
if (mapping_id == std::end(...)) {
  return default;
}
return *mapping_it;
```

hikinggrass replied:
I guess something like this would be quite readable:
```cpp
const auto& mappings = module_tier_mappings.value();
const auto mapping_it = mappings.implementations.find(impl_id);
if (mapping_it == mappings.implementations.end()) {
    return mappings.module;
}
const auto& [key, mapping] = (*mapping_it);
return mapping;
```

The alternative doesn't quite feel as readable:
```cpp
...
return (*mapping_it).second;
```
Or what do you think?

a-w50 commented on lib/everest.cpp:
```diff
@@ -158,30 +158,33 @@ Everest::Everest(std::string module_id_, const Config& config_, bool validate_da
+            std::string interface_name = this->module_manifest.at("requires").at(req.id).at("interface");
+            json interface_def = this->config.get_interface_definition(interface_name);
+            std::list<std::string> allowed_error_types;
+            for (const auto& error_namespace_it : interface_def["errors"].items()) {
```
these changes are probably not necessary, if you keep `get_requirements`

hikinggrass replied:
I've re-added get_requirements, that does lead to some code duplication though since information required for the Fulfillments is not available from this function

a-w50 commented on lib/everest.cpp:
```diff
@@ -252,6 +255,22 @@ void Everest::register_on_ready_handler(const std::function<void()>& handler) {
+        // if no specific implementation mapping is given, use the module mapping
+        return mapping.module;
+    }
+    return mapping.implementations.at(impl_id);
```
Same story with the iterator as above.  Function should be const and could also return by `const ref&` (for this you would need to have a `static std::optional<Mapping> default_value {std::nullopt}` which you would return, otherwise you directly return the references instead of unecessary copying).

a-w50 reviewed (changes requested):
See the comments, please address function signatures and if possible remove the mapping from the `Requirement` struct.

hikinggrass committed:
Remove get_imple_mapping from ModuleAdapter

Functionality now provided via a free helper function with the same name

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move mapping member from Requirement to Fulfilment

re-introduce get_requirements function

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Rename function so it doesn't share a name with another one returning a different type

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Make more mapping related variables&functions const

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add reworded comment

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 commented on lib/config.cpp:
```diff
@@ -710,9 +710,9 @@ std::list<Requirement> Config::get_requirements(const std::string& module_id) co
     std::list<Requirement> res;
 
-    std::string module_name = get_module_name(module_id);
+    const std::string module_name = get_module_name(module_id);
```
could be `auto`

a-w50 commented on lib/config.cpp:
```diff
@@ -710,9 +710,9 @@ std::list<Requirement> Config::get_requirements(const std::string& module_id) co
+    const std::string module_name = get_module_name(module_id);
     for (const std::string& req_id : Config::keys(this->manifests.at(module_name).at("requires"))) {
-        json resolved_req = this->resolve_requirement(module_id, req_id);
+        const json resolved_req = this->resolve_requirement(module_id, req_id);
```
could be `auto`

a-w50 commented on include/utils/types.hpp:
```diff
@@ -104,11 +108,42 @@ struct ModuleTierMappings {
+    std::string module_id;
+    std::string implementation_id;
+    Requirement requirement;
+    std::optional<Mapping> mapping;
```
Looking at the `everestpy` code, a `Fulfillment` only contains the module id and implementation id.  I should probably have named that `FulfillmentID` back then (and similar `RequirementID`).
What I've said about putting the mapping into the `Requirement` in the previous review, also applies to the `Fulfillment` here.  `Fulfillment` and `Requirement` should only be identifiers.  Their connection should be represented as `Requirement -> Fulfillment` (as is done in everestpy).
What should be done here, is to have a function like `resolve_requirements(module id)` which returns a map (or a vector) of type `std::map<Requirement, Fulfillment>` (this is actually the logic/code shared by `get_requirements` and `get_fulfillments` - both functions should use this common function then).
Having this map, you would then need to transform it:
- in case of `get_requirements` you only need to return the `keys`
- in case of `get_fulfillments`, this should be the identity, i.e. just returning the connections

Now you still don't have access to the mapping.  In order to expose that, you'll need to do a transform of the above mentioned `std::map`, where you iterate over the keys (requirements) and transform the fulfillments into their corresponding mapping, i.e. a map like `std::map<Requirement, Mapping>`.  And with this you should proceed then. This function could be called `transform_to_mapping`, ideally a free function which takes a functor (std::function), which can retrieve a mapping from a fulfillment.

The basic idea behind this separation is to have identifier structs (like Requirement or Fulfillment).  Now we may have different parts in the code which want to attach information to these identifiers (like attaching a Fulfillment to a Requirement, or attaching a Mapping to a Fulfillment or Requirement).  If we would put all that stuff into the Requirement or Fulfillment identifiers, we create coupling between these sections, and we should avoid that.  We achieve better separation, when we create new or custom types (like an additional map), which each contain specific mappings.

a-w50 commented on lib/runtime.cpp:
```diff
@@ -495,11 +497,12 @@ int ModuleLoader::initialize() {
         // FIXME (aw): would be nice to move this config related thing toward the module_init function
-        std::vector<cmd> cmds =
-            this->callbacks.everest_register(config.get_main_config()[this->module_id]["connections"]);
+        std::vector<cmd> cmds = this->callbacks.everest_register(config.get_fulfillments(this->module_id));
```
I just saw that now, why do we pass the fulfillments via `everest_register`?  `everest_register` wasn't intended to transfer this kind of information to the module - it's intend was to get the implemented commands of the module (if I remember correctly).
This information should be passed via the init function.  If possible, I would add this information to the `ModuleInfo` struct, in case it can be consumed directly via user code, or if not pass it as an additional argument and then it's consumption needs to take place in the `ld-ev` code which finds the corresponding requirement and sets its `_mapping` from that.

hikinggrass committed:
Add resolve_requirements that is used as a basis for refactored get_requirements and get_fulfillments

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove code that isn't used

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Introduce RequirementInitializer and RequirementInitialization

These can be used in (generated) user code to initialize requirements with their fulfillments and optional mappings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove redundant mapping from Fulfillment

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Turn operator< of Requirement into free function

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

a-w50 commented on include/utils/types.hpp:
```diff
@@ -132,20 +132,29 @@ struct TelemetryConfig {
 struct Fulfillment {
     std::string module_id;
     std::string implementation_id;
     Requirement requirement;
```
Is there any reason, why requirement is still here?

hikinggrass replied:
It's actively used in everestpy and I haven't looked into how to replace it yet, might do that though

a-w50 replied:
I see, sorry I wasn't aware of that

a-w50 commented on include/utils/types.hpp:
```diff
@@ -132,20 +132,29 @@ struct TelemetryConfig {
     std::optional<Mapping> mapping;
 };
 
+using RequirementInitialization = std::map<std::string, std::vector<RequirementInitializer>>;
```
I'm uncertain with that name:  `RequirementInitialization` sounds more like the result of the initialization of a requirement than a mapping from each requirement name to its initializer.

hikinggrass replied:
I'm open for suggestions :smile: that was the best I could come up with at the time :sweat_smile:

a-w50 commented on lib/config.cpp:
```diff
@@ -705,25 +705,40 @@ json Config::resolve_requirement(const std::string& module_id, const std::string
         const json resolved_req = this->resolve_requirement(module_id, req_id);
         if (!resolved_req.is_array()) {
+            const auto resolved_module_id = resolved_req.at("module_id");
+            const auto resolved_impl_id = resolved_req.at("implementation_id");
```
These should be `auto&`, right?

a-w50 commented on lib/config.cpp:
```diff
@@ -705,25 +705,40 @@ json Config::resolve_requirement(const std::string& module_id, const std::string
+
+    for (const auto& [requirement, fulfillment] : this->resolve_requirements(module_id)) {
+        res.push_back(requirement);
+        (void)fulfillment; // fulfillment is not used here
```
This void is a not that beautiful.  `[[maybe_unused]]` could be used?  Or just normal `it->first`?

a-w50 commented on lib/config.cpp:
```diff
@@ -732,26 +747,21 @@ std::map<std::string, std::vector<Fulfillment>> Config::get_fulfillments(const s
+    return res;
+}
+
+RequirementInitialization Config::get_requirement_initialization(const std::string& module_id) const {
```
This function is now _very_ readable :+1:

a-w50 commented on lib/config.cpp:
```diff
@@ -732,26 +747,21 @@ std::map<std::string, std::vector<Fulfillment>> Config::get_fulfillments(const s
+
+    for (const auto& [requirement, fulfillment] : this->resolve_requirements(module_id)) {
+        const auto& mapping = this->get_3_tier_model_mapping(fulfillment.module_id, fulfillment.implementation_id);
+        res[requirement.id].push_back({requirement, fulfillment, mapping});
```
This is a bit tricky. We implicitly rely on the fact, that requirements are sorted within the map according to their index.  This should be guaranteed, but still is not obvious to the reader.  Having the requirement without it's index, but instead pointing at a vector of `Fulfillments` would probably make the code even simpler.

a-w50 commented on include/utils/types.hpp:
```diff
@@ -132,20 +132,29 @@ struct TelemetryConfig {
 struct Requirement {
     Requirement() = default;
     Requirement(const std::string& requirement_id_, size_t index_);
```
These could be dropped now

hikinggrass replied:
Ah I thought I removed them already, must have overlooked that

hikinggrass committed:
Remove constructors of Requirement

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove ugly void cast

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
More const auto(&) usage

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix everestrs build after removal of Requirement constructor

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge branch 'main' into feature/expose-mappings

hikinggrass committed:
Restructure mappings in config schema.

Now all mappings are defined under a "mapping" key that contains a "module" and/or "implementations" key which in turn can have mappings for individual implementations defined

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>
